### PR TITLE
Session cross-reference / transaction mismatch

### LINKED ISSUE #20855: Session cross-reference / transaction mismatch
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

#20793 introduced a memory leak similar to #19396.

Context contains objects from session `S1`.
Session `S1` contains objects that refer to transaction `T1`, and often to session `S1` (see e.g. [`RealmAdapter`](https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java#L53)).

Now in the new variant of `runJobInTransactionWithResult`, a new session `S2` is created with transaction `T2`.
It however refers objects from `S1`, and until that transaction is completed, it prevents its garbage collection.

Furthermore, the objects in the new session `S2` context are controlled by transaction `T1`, while they should be controlled by transaction `T2`.

### Version

main

### Expected behavior

Objects from a transaction of a session are controlled by that transaction.

### Actual behavior

Objects from the context of a session are controlled by parent transaction.

### How to Reproduce?

Leverage new variant of `runJobInTransactionWithResult`.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rmartinc**
Closes https://github.com/keycloak/keycloak/issues/20855

@hmlnarik Sending the PR. All CI passed in my branch.

Now I thought that another solution would be re-loading context models lazily instead of pro-actively. For example for realm, the `DefaultKeycloakContext` can be initialized just with the `realmId` and if `getRealm` is called the context gets the realm via the String property (more complicated for the rest of models, but same idea). This would allow to use any model in the context, and limitations in DBs are just exposed when using those DBs. But as we decided to use this option I'm just sending it this way. Maybe I'm overthinking this issue, but I just don't see a good solution for it. :confused:

Thanks!

### CODE REVIEW

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 1]
```diff
@@ -275,14 +277,72 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
...
+     */
+    private static void cloneContextToSession(final KeycloakContext origContext, final KeycloakSession targetSession,
+            final boolean includeAuthenticatedSessionModel) {
+        if (origContext == null) {
+            return;
+        }
+
+        // setup realm model if necessary.
+        if (origContext.getRealm() == null) {
+            return;
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:313]: Please address https://github.com/rmartinc/keycloak/commit/a7eafec4da46d97a6fa88c7eae31e6113508a941#r117140224

**Review by hmlnarik — CHANGES_REQUESTED** [review]
This PR leads to having two places where the same solution is implemented slightly differently. Is there any reason for it?

I suggest reusing https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/common/util/ResponseSessionTask.java#L92-L116 for the `cloneContextToSession` body, and make `ResponseSessionTask.setupSessionContext` call this utility function to keep the logic in one place.

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java (hunk 1 above)
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java]: Done! Now the three model objects are read dependently.

**Comment by rmartinc** [comment]
> This PR leads to having two places where the same solution is implemented slightly differently. Is there any reason for it?

Yes, I copied the method but if you check it the previous impl didn't allow realmModel null and clientModel not null. If you see the clientModel is read using the previous read realmModel ([here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/common/util/ResponseSessionTask.java#L104)). So, if the realmModel was null a NPE was thrown here. That's way I understood that one model object required the previous one.

> I suggest reusing https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/common/util/ResponseSessionTask.java#L92-L116 for the `cloneContextToSession` body, and make `ResponseSessionTask.setupSessionContext` call this utility function to keep the logic in one place.

I'm changing it a bit to avoid the NPE and read a non-null client with a null realm.

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 2]
```diff
@@ -275,14 +277,74 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
...
+        if (origContext.getRealm() != null) {
+            RealmModel realmModel = targetSession.realms().getRealm(origContext.getRealm().getId());
+            if (realmModel != null) {
+                targetSession.getContext().setRealm(realmModel);
+            }
+        }
+
+        // setup client model if necessary.
+        if (origContext.getClient() != null) {
+            RealmModel realmModel = targetSession.realms().getRealm(origContext.getClient().getRealm().getId());
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:321]: Last bit, double lookup should be avoided: value from line 313 can be reused if `origContext.getRealm().getId() == origContext.getClient().getRealm().getId()`.

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 3]
```diff
@@ -275,14 +277,74 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
...
+                ClientModel clientModel = targetSession.clients().getClientById(realmModel, origContext.getClient().getId());
+                if (clientModel != null) {
+                    targetSession.getContext().setClient(clientModel);
+                }
+            }
+        }
+
+        // setup auth session model if necessary.
+        if (includeAuthenticatedSessionModel && origContext.getAuthenticationSession() != null) {
+            RealmModel realmModel = targetSession.realms().getRealm(origContext.getAuthenticationSession().getRealm().getId());
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:332]: Same here

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java [hunk 4]
```diff
@@ -275,14 +277,74 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakC
...
+            if (realmModel != null) {
+                ClientModel clientModel = targetSession.clients().getClientById(realmModel, origContext.getAuthenticationSession().getClient().getId());
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:334]: and here.

**Review by hmlnarik — COMMENTED** [review]
Thank you, there is one last bit and will be good to go: Most of the times the client will be from the same realm as set in the context.realm, and similarly with the auth sessions. The call to retrieve object by ID is costly especially when cache is disabled. Please see comments inline and reuse objects as much as possible

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java (hunk 2 above)
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java]: Done! I think the test failures are spurious (CI passed with the same commit in my branch: https://github.com/rmartinc/keycloak/actions/runs/5228904068).

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java (hunk 4 above)
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java]: Done!

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java (hunk 3 above)
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java]: Done!

**Review by hmlnarik — APPROVED** [review]
Looks good, thank you!

**Pull request merged**